File: services/voice-agent/src/voice_agent/twilio/call_credential.py
"""Verifies M7 call credentials minted by apps/api (Twilio inbound calls).

Mirrors apps/api's src/auth/call-credential.ts algorithm exactly (HMAC-SHA256
over a base64url-encoded JSON payload), using only stdlib
hmac/hashlib/json/base64 -- same "no new dependency for one small,
well-documented algorithm" reasoning as twilio/signature.py. See the
approved M7 plan sections 6 and 17.

voice-agent never mints a call credential -- only apps/api does, at the
Twilio phone-number-lookup internal endpoint. This module is verify-only,
used by routes/twilio.py's media-stream WebSocket handler to establish
organization identity LOCALLY, before session.run_session() is ever
called. The organization_id returned by a successful verify() call here --
never a separate, unauthenticated WebSocket parameter -- is the sole
source of truth for which organization a session belongs to; see the
approved M7 plan section 2.

The payload contract is fixed:
{"organization_id": "<uuid>", "call_sid": "<Twilio CallSid>",
"exp": <unix_timestamp_int>} -- exp must be an actual int, not a float and
not a bool (Python's bool is a subclass of int, so isinstance(x, int)
alone would silently accept True/False as 1/0; both checks are explicit
below).

Node's base64url encoding omits padding; Python's base64.urlsafe_b64encode
includes it and urlsafe_b64decode requires correctly padded input -- this
module adds/strips padding explicitly so tokens minted by the Node
implementation verify correctly here. See
tests/test_twilio_call_credential.py's interoperability test, which mints
a token using an independent reimplementation of the Node algorithm and
verifies it with this module.

Every exception below carries only a short, fixed, safe-to-log message
(never the token or its claims) -- but callers should still log only
type(exc).__name__, matching this codebase's established
never-log-raw-exception-text discipline (see session.py, logging_config.py).
"""

from __future__ import annotations

import base64
import hashlib
import hmac
import json
import time
from dataclasses import dataclass
from typing import Any


class CallCredentialError(Exception):
    """Base class for all call-credential verification failures."""


class MalformedCallCredentialError(CallCredentialError):
    """Not two dot-separated parts, invalid base64, invalid JSON, or a
    payload missing/mistyping a required claim (including exp being a
    float or a bool instead of a plain int)."""


class InvalidCallCredentialSignatureError(CallCredentialError):
    """The HMAC signature does not match -- wrong secret, or the payload
    was tampered with after signing."""


class ExpiredCallCredentialError(CallCredentialError):
    """The credential's exp claim is in the past."""


class CallSidMismatchError(CallCredentialError):
    """The credential's call_sid claim does not match the real Twilio
    call_sid for the connection presenting it."""


@dataclass(frozen=True)
class VerifiedCallCredential:
    """The two claims voice-agent actually needs, extracted only once
    verification has fully succeeded."""

    organization_id: str
    call_sid: str


def _b64url_encode_no_pad(data: bytes) -> str:
    return base64.urlsafe_b64encode(data).decode("ascii").rstrip("=")


def _b64url_decode_no_pad(value: str) -> bytes:
    padding = "=" * (-len(value) % 4)
    return base64.urlsafe_b64decode(value + padding)


def verify_call_credential(
    token: str,
    secret: str,
    expected_call_sid: str,
) -> VerifiedCallCredential:
    """Verifies a call credential minted by apps/api's generateCallCredential.

    Raises a specific CallCredentialError subclass on any failure -- there
    is no partial-success case, matching apps/api's own verifyCallCredential.
    `expected_call_sid` must be the real Twilio call_sid for the connection
    presenting this token (from parse_telephony_websocket()'s call_data,
    never from a value the token itself claims to be) -- this is what
    scopes a credential to one specific call rather than being replayable
    against any future connection.
    """
    parts = token.split(".")
    if len(parts) != 2:
        raise MalformedCallCredentialError("malformed")
    payload_b64, sig_b64 = parts

    expected_sig_b64 = _b64url_encode_no_pad(
        hmac.new(secret.encode("utf-8"), payload_b64.encode("utf-8"), hashlib.sha256).digest()
    )
    if not hmac.compare_digest(expected_sig_b64, sig_b64):
        raise InvalidCallCredentialSignatureError("signature")

    try:
        payload: Any = json.loads(_b64url_decode_no_pad(payload_b64))
    except (ValueError, UnicodeDecodeError) as exc:
        raise MalformedCallCredentialError("malformed") from exc

    # exp must be an actual int -- isinstance(x, int) alone would also
    # accept a bool (True/False are ints in Python) or would need a
    # separate float check, so both are explicit and separate here.
    if (
        not isinstance(payload, dict)
        or not isinstance(payload.get("organization_id"), str)
        or not isinstance(payload.get("call_sid"), str)
        or not isinstance(payload.get("exp"), int)
        or isinstance(payload.get("exp"), bool)
    ):
        raise MalformedCallCredentialError("malformed")

    if payload["call_sid"] != expected_call_sid:
        raise CallSidMismatchError("call_mismatch")

    if payload["exp"] < time.time():
        raise ExpiredCallCredentialError("expired")

    return VerifiedCallCredential(
        organization_id=payload["organization_id"],
        call_sid=payload["call_sid"],
    )

File: apps/api/src/auth/call-credential.ts
import { createHmac } from "node:crypto";
import { safeCompare } from "../middleware/require-service-auth.js";

/**
 * M7: a short-lived, cryptographically bound credential minted by apps/api
 * for exactly one inbound Twilio call, used in place of the long-lived M5
 * organization service token (organization-service-token.ts) for that one
 * call. apps/api is the only minter; voice-agent verifies it independently
 * before ever calling session.run_session() (see
 * services/voice-agent/src/voice_agent/twilio/call_credential.py), and
 * apps/api's own requireOrganizationAuth middleware re-verifies it again on
 * every internal-API call — see middleware/require-organization-auth.ts.
 *
 * Deliberately NOT a JWT: a two-part `payload.signature` token, one
 * hard-coded algorithm (HMAC-SHA256), no "alg" field for an attacker to
 * manipulate, no new dependency. See the approved M7 plan §6.
 *
 * Payload shape (JSON, snake_case — must match the Python verifier's
 * expectations byte-for-byte):
 *   { organization_id: string, call_sid: string, exp: number (unix secs) }
 */

/** 4 hours — see the approved M7 plan §3 for the full justification: real
 * calls are realistically minutes long, so this is never at risk of
 * expiring mid-call; call_sid binding (not the TTL) is the primary defense
 * against replay across unrelated calls. */
export const DEFAULT_CALL_CREDENTIAL_TTL_SECONDS = 4 * 60 * 60;

interface CallCredentialPayload {
  organization_id: string;
  call_sid: string;
  exp: number;
}

export type CallCredentialVerifyResult =
  | { valid: true; callSid: string }
  | {
      valid: false;
      reason: "malformed" | "signature" | "expired" | "org_mismatch" | "call_mismatch";
    };

function sign(payloadB64: string, secret: string): string {
  return createHmac("sha256", secret).update(payloadB64).digest("base64url");
}

/**
 * Mints a call credential bound to exactly one organization and one Twilio
 * call. Called exactly once per inbound call, from the Twilio voice
 * webhook handler, only after the webhook's signature has already been
 * validated (see the approved M7 plan §12's strict fail-closed order).
 */
export function generateCallCredential(
  organizationId: string,
  callSid: string,
  secret: string,
  ttlSeconds: number = DEFAULT_CALL_CREDENTIAL_TTL_SECONDS,
): string {
  const payload: CallCredentialPayload = {
    organization_id: organizationId,
    call_sid: callSid,
    exp: Math.floor(Date.now() / 1000) + ttlSeconds,
  };
  const payloadB64 = Buffer.from(JSON.stringify(payload), "utf8").toString("base64url");
  const sigB64 = sign(payloadB64, secret);
  return `${payloadB64}.${sigB64}`;
}

/**
 * Verifies a call credential against a known organization id (the caller —
 * requireOrganizationAuth — already knows which :organizationId the URL
 * names; this checks the credential actually authorizes that exact org).
 * `expectedCallSid`, when provided, additionally requires the credential's
 * own call_sid claim to match it — used where the real Twilio call_sid is
 * independently known (the WS entry point; apps/api's HTTP requests don't
 * carry one, so requireOrganizationAuth omits it).
 *
 * Any failure mode — malformed structure, bad signature, wrong org, wrong
 * call_sid, or expired — is reported as `{ valid: false, reason }`; the
 * reason is a fixed enum-like string safe to log, never the token or
 * payload contents themselves (see the approved M7 plan §6).
 */
export function verifyCallCredential(
  token: string,
  organizationId: string,
  secret: string,
  expectedCallSid?: string,
): CallCredentialVerifyResult {
  const parts = token.split(".");
  if (parts.length !== 2) {
    return { valid: false, reason: "malformed" };
  }
  const [payloadB64, sigB64] = parts as [string, string];

  const expectedSig = sign(payloadB64, secret);
  if (!safeCompare(sigB64, expectedSig)) {
    return { valid: false, reason: "signature" };
  }

  let payload: CallCredentialPayload;
  try {
    const decoded = JSON.parse(Buffer.from(payloadB64, "base64url").toString("utf8")) as unknown;
    if (
      typeof decoded !== "object" ||
      decoded === null ||
      typeof (decoded as CallCredentialPayload).organization_id !== "string" ||
      typeof (decoded as CallCredentialPayload).call_sid !== "string" ||
      typeof (decoded as CallCredentialPayload).exp !== "number"
    ) {
      return { valid: false, reason: "malformed" };
    }
    payload = decoded as CallCredentialPayload;
  } catch {
    return { valid: false, reason: "malformed" };
  }

  if (payload.organization_id !== organizationId) {
    return { valid: false, reason: "org_mismatch" };
  }
  if (expectedCallSid !== undefined && payload.call_sid !== expectedCallSid) {
    return { valid: false, reason: "call_mismatch" };
  }
  if (payload.exp < Math.floor(Date.now() / 1000)) {
    return { valid: false, reason: "expired" };
  }

  return { valid: true, callSid: payload.call_sid };
}
